### PR TITLE
Adding support for optional parameters

### PR DESCRIPTION
**Pull request by scottfrazer**

### CODE REVIEW

File: src/main/scala/cromwell/binding/Call.scala [hunk 1]
```diff
@@ -31,13 +31,16 @@ case class Call(alias: Option[String], taskFqn: FullyQualifiedName, task: Task,
     else throw new UnsupportedOperationException("parent is write-once")
   }
 
+  private def unsatisfiedTaskInputs: Seq[TaskInput] = task.inputs.filterNot {case i => inputMappings.contains(i.name)}
   /**
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/Call.scala]: newline

File: src/main/scala/cromwell/binding/Call.scala [hunk 2]
```diff
@@ -31,13 +31,16 @@ case class Call(alias: Option[String], taskFqn: FullyQualifiedName, task: Task,
...
    * Map of task-local input names to type for every input that this
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/Call.scala]: I'm not sure what this comment is saying, but I'm pretty sure it's not what I thought this function was doing. Can you rephrase the purpose of this function in long form here in the github commentary?

File: src/main/scala/cromwell/binding/TaskInput.scala [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package cromwell.binding
+
+import cromwell.binding.types.WdlType
+
+case class TaskInput(name: String, types: Seq[WdlType], postfixQuantifier: Option[String])
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/TaskInput.scala]: Can you explain why a task input would have multiple types?

File: src/main/scala/cromwell/binding/WdlNamespace.scala [hunk 4]
```diff
@@ -266,25 +302,37 @@ case class WdlNamespace(ast: Ast, source: WdlSource, importResolver: ImportResol
...
-      case _ if rawInputs.contains(fqn) =>
-        val rawInput = rawInputs.get(fqn).get
-        wdlType.coerceRawValue(rawInput).recoverWith {
-          case e => Failure(new UnsatisfiedInputsException(s"Failed to coerce input $fqn value $rawInput of ${rawInput.getClass} to $wdlType."))
+    def coerceRawInput(input: WorkflowInput): Try[Option[WdlValue]] = input.fqn match {
+      case _ if rawInputs.contains(input.fqn) =>
+        /* Workflow inputs can sometimes have multiple types.  For example, `${sep="," var+}`
+         * can accept a `String` or `Array[String]`.  This coercion tries to coerce to ALL valid types
+         * and just chooses the first one that works, or returns a Failure.
+         */
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: Can we know which of those types is correct at WDL compile time or does it need to happen at WDL runtime (i.e. workflow processing)? 

What I mean is, is the absolute best we can do here to say "yeah, something works" and then hope for the best later on?

File: src/main/scala/cromwell/binding/command/ParameterCommandPart.scala [hunk 5]
```diff
@@ -2,16 +2,20 @@ package cromwell.binding.command
 
 import cromwell.binding.WdlExpression
 import cromwell.binding.types.WdlType
-import cromwell.binding.values.{WdlPrimitive, WdlValue}
+import cromwell.binding.values.{WdlString, WdlArray, WdlPrimitive, WdlValue}
 import scala.util.Success
 
-case class ParameterCommandPart(wdlType: WdlType, name: String) extends CommandPart {
+case class ParameterCommandPart(wdlType: WdlType, name: String,
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/command/ParameterCommandPart.scala]: Can you add commentary as to what the new params are? (in code, not here)

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package cromwell.binding.types
+
+import cromwell.binding.values.{WdlInteger, WdlArray}
+import spray.json.JsArray
+
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: Is this part of optional parameters or separate? (the latter is fine IMO, just for my own understanding)

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+case class WdlArrayType(memberType: WdlType) extends WdlType {
+  val toWdlString: String = s"Array[${memberType.toWdlString}]"
+
+  private def coerceIterable(s: Seq[Any]): WdlArray = {
+    val coerced = s.map {e => memberType.coerceRawValue(e).get}
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: Use the `_` instead of the `e`

File: src/main/scala/cromwell/binding/WdlSyntaxErrorFormatter.scala [hunk 8]
```diff
@@ -158,4 +158,25 @@ case class WdlSyntaxErrorFormatter(terminalMap: Map[Terminal, WdlSource]) extend
...
+    s"""ERROR: Map type should have two parameterized types (line ${mapDecl.getLine}, col ${mapDecl.getColumn}):
+     |
+     |${pointToSource(mapDecl)}
+     """.stripMargin
+  }
+
+  def mapKeyMustBeAPrimitive(mapDecl: Terminal): String = {
+    s"""ERROR: Map type must have a primitive as its key (line ${mapDecl.getLine}, col ${mapDecl.getColumn}):
+                                                                                                                |
+                                                                                                                |${pointToSource(mapDecl)}
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlSyntaxErrorFormatter.scala]: indent fail

File: src/main/scala/cromwell/binding/WorkflowInput.scala [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package cromwell.binding
+
+import cromwell.binding.types.WdlType
+
+case class WorkflowInput(fqn: FullyQualifiedName, types: Seq[WdlType], postfixQuantifier: Option[String]) {
+  val optional = postfixQuantifier match {
+    case Some(s) => Set("?", "*").contains(s)
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/WorkflowInput.scala]: These quantifier `Set`s should be constants

File: src/main/scala/cromwell/binding/types/WdlMapType.scala [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package cromwell.binding.types
+
+case class WdlMapType(keyType: WdlPrimitiveType, valueType: WdlType) extends WdlType {
+  val toWdlString: String = s"Map[${keyType.toWdlString}, ${valueType.toWdlString}]"
+  override protected def coercion = ???
+  override def fromRawString(s: String) = ???
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/types/WdlMapType.scala]: won't this need to be implemented for persistence?

File: src/main/scala/cromwell/engine/workflow/WorkflowActor.scala [hunk 11]
```diff
@@ -198,8 +198,19 @@ case class WorkflowActor(workflow: WorkflowDescriptor,
       Await.ready(futureCallsAndInputs, DatabaseTimeout)
       futureCallsAndInputs.value.get match {
         case Success(callsAndInputs) =>
-          log.info(s"$tag calls and inputs are " + callsAndInputs)
-          callsAndInputs foreach { case (call, inputs) => startActor(call, inputs)}
+          callsAndInputs foreach { case (call, inputs) =>
+
+            if (inputs.nonEmpty) {
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: can you pull the logging out to a method?

File: src/test/scala/cromwell/ThreeStepActorSpec.scala [hunk 12]
```diff
@@ -1,87 +1,51 @@
...
                     |  docker: "ubuntu:latest"
                     |}
                   """.stripMargin,
-        expectations = CannedExpectations)
+        expectedOutputs = CannedExpectations)
     }
 
     "pass files properly" taggedAs DockerTest in {
-      runAndAssertCorrectness(sampleWdl = SampleWdl.FilePassingThreeStep)
+      val event = EventFilter.info(pattern = s"starting calls: three_step.cgrep, three_step.wc", occurrences = 1)
```
Comment (mcovarr) [inline src/test/scala/cromwell/ThreeStepActorSpec.scala]: what's this for?

File: src/main/scala/cromwell/binding/WdlNamespace.scala [hunk 13]
```diff
@@ -216,7 +216,22 @@ case class WdlNamespace(ast: Ast, source: WdlSource, importResolver: ImportResol
   private def processCommandParameter(ast: Ast): ParameterCommandPart = {
     val wdlType = processWdlType(ast.getAttribute("type"))
     val name = ast.getAttribute("name").asInstanceOf[Terminal].getSourceString
-    new ParameterCommandPart(wdlType, name)
+    val prefix = ast.getAttribute("prefix") match {
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: Option / map?

File: src/main/scala/cromwell/binding/WdlNamespace.scala (hunk 4 above)
Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: IMHO + should coerce to `Array[String]` in all cases

File: src/test/scala/cromwell/ThreeStepActorSpec.scala [hunk 14]
```diff
@@ -1,87 +1,51 @@
...
-      runAndAssertCorrectness(sampleWdl = SampleWdl.CannedThreeStep, expectations = CannedExpectations)
+      runWdlAndAssertOutputs(
+        sampleWdl = SampleWdl.CannedThreeStep,
+        EventFilter.info(pattern = eventMessage, occurrences = 1),
+        expectedOutputs = CannedExpectations)
     }
   }
 
   "A Dockerized three step workflow" should {
+    val event = EventFilter.info(pattern = s"starting calls: three_step.cgrep, three_step.wc", occurrences = 1)
```
Comment (mcovarr) [inline src/test/scala/cromwell/ThreeStepActorSpec.scala]: also appears unused

File: src/main/scala/cromwell/binding/Call.scala (hunk 2 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/Call.scala]: Yes I can see how that is confusing.  The comment is a little outdated.  Should probably read:

> Returns a Seq[WorkflowInput] representing the inputs to the call that are needed before its command can be constructed.  This excludes inputs that are satisfied via the 'input' section of the Call definition.

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/Call.scala]: Unless you want to change wording I'll add that as the comment.

File: src/main/scala/cromwell/binding/TaskInput.scala (hunk 3 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/TaskInput.scala]: I'll add this as the comment for `types` in the ScalaDoc

> Because tasks define inputs that take 0-or-more and 1-or-more parameters (e.g. `${sep="," var*}`), the corresponding acceptable types would be `WdlString` and `Array[WdlString]`.  `WdlString` would effectively be treated as an `Array[WdlString]` of length 1.  When expanding this particular command line parameter, the items would be concatenated together using the separator in between them.  Allowing both `WdlString` and `Array[WdlString]` is a convenience mostly, so that client's don't have to specify arrays of size one.  If this ends up adding a lot of code complexity, then `types` can be changed to accept `WdlType` as its type without affecting too much

File: src/main/scala/cromwell/binding/WdlNamespace.scala (hunk 4 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: @mcovarr I'm not opposed to that idea.  The only argument I have for keeping it is that it meets user expectations a little closer.  For example:

```
${sep=" " File x+}
```

If we're telling users that `x` only accepts an `Array[File]` then I think that's misleading because the type really is `File`.  The language I've always used when describing this in english is: "Treated as a regular parameter if you pass it a matching primitive type.  If given an array of that primitive type, it will expand it via the separator"

I'm trying to brainstorm some alternatives to the WDL syntax that appease both concerns:

```
${sep=" " Array[File] x}
```

Then, the rule can be "if `sep` is specified, append the elements in the array together into a string.  If not, then replace this variable with a `File` reference to the array values."

Seems okay.  Then we'd be getting rid of `?` and `*` altogether, which does simplify things a bit.  The current approach offers slightly better static checking because it *requires* `sep` to be specified if `?` or `*` are specified.

What do you think?

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: @geoffjentry We could only know at runtime because it depends on what the user passes as the value for that variable in their JSON file.  See discussion above though, if we change the spec we can make `WorkflowInput` only accept one type, which would certainly simplify the next 4 lines a bit.  Let me know what you think of my prior comment?

File: src/main/scala/cromwell/binding/WdlSyntaxErrorFormatter.scala (hunk 8 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlSyntaxErrorFormatter.scala]: IntelliJ is really really smart about some things and very strangely incompetent at other things... pasting margin'd code is not its strong suite.

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala (hunk 6 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: Let me double check but I think since `*` and `+` utilize the concept of Arrays a little bit I think that's why it's in here, but that may not actually be true.

File: src/main/scala/cromwell/binding/types/WdlMapType.scala (hunk 10 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/types/WdlMapType.scala]: We will, I just didn't implement it yet.  I think I just added this for completeness but I should probably remove it and make it part of another PR.

File: src/test/scala/cromwell/ThreeStepActorSpec.scala (hunk 14 above)
Comment (scottfrazer) [inline src/test/scala/cromwell/ThreeStepActorSpec.scala]: Bleh, yeah, code cruft...

File: src/main/scala/cromwell/binding/WdlNamespace.scala (hunk 13 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: I don't understand

Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]:
```
Option(ast.getAttribute("prefix")) map { case t: Terminal => t.sourceString  }
```

File: src/main/scala/cromwell/binding/WdlNamespace.scala (hunk 4 above)
Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: If an expression generates a singular value for a + or * quantified parameter that seems okay, couldn't we just automatically put it in a collection of size 1?  I was not proposing (nor would I want) to expose the `Array[File]` in WDL. :smile:  I think the various quantifiers are semantically useful and also wasn't looking to remove them (specifically, more than one element for a ? quantified parameter should be an error).

File: src/main/scala/cromwell/binding/WdlNamespace.scala [hunk 15]
```diff
@@ -266,25 +292,37 @@ case class WdlNamespace(ast: Ast, source: WdlSource, importResolver: ImportResol
...
+    }.toMap
 
     val (successes, failures) = tryCoercedValues.partition { case (_, tryValue) => tryValue.isSuccess }
     if (failures.isEmpty) {
-      Try(successes.map { case (key, tryValue) => key -> tryValue.get })
+      Try(successes.map {
+        case (key, tryValue) => key -> tryValue.get
+      }.filter {y => y._2.isDefined}.map {
+        case (key, optionValue) => key -> optionValue.get
+      })
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: Not critical, but more clear I think:

```
for {
  (key, tryValue) <- successes
  optionValue = tryValue.get if tryValue.get.isDefined
} yield (key, optionValue.get)
```

File: src/test/scala/cromwell/CromwellTestkitSpec.scala [hunk 16]
```diff
@@ -72,4 +89,31 @@ with DefaultTimeout with ImplicitSender with WordSpecLike with Matchers with Bef
...
+    val coercedInputs = namespace.coerceRawInputs(sampleWdl.rawInputs).get
+    val descriptor = WorkflowDescriptor(UUID.randomUUID(), namespace, sampleWdl.wdlSource(runtime), sampleWdl.wdlJson, coercedInputs)
+    TestFSMRef(new WorkflowActor(descriptor, new LocalBackend, dataAccess))
+  }
+
+  def runWdlAndAssertOutputs(sampleWdl: SampleWdl, event: EventFilter, runtime: String = "", expectedOutputs: Map[FullyQualifiedName, WdlValue] = Map.empty): Unit = {
+    val fsm = buildFsmWorkflowActor(sampleWdl, runtime)
+    assert(fsm.stateName == WorkflowSubmitted)
+
+    event.intercept {
```
Comment (mcovarr) [inline src/test/scala/cromwell/CromwellTestkitSpec.scala]: `event` is more correctly an `eventFilter`

File: src/test/scala/cromwell/ThreeStepActorSpec.scala [hunk 17]
```diff
@@ -1,87 +1,53 @@
...
-
 object ThreeStepActorSpec {
-  val TestExecutionTimeout = 5000 milliseconds
   val CannedExpectations = Map(
-    "three_step.cgrep.count" -> 3,
-    "three_step.wc.count" -> 6)
+    "three_step.cgrep.count" -> WdlInteger(3),
+    "three_step.wc.count" -> WdlInteger(6)
+  )
+  val eventMessage = s"starting calls: three_step.cgrep, three_step.wc"
```
Comment (mcovarr) [inline src/test/scala/cromwell/ThreeStepActorSpec.scala]: constants are capitalized!  :smile:

File: src/main/scala/cromwell/binding/TaskInput.scala (hunk 3 above)
Comment (mcovarr) [inline src/main/scala/cromwell/binding/TaskInput.scala]: can postfix quantifier reasonably default to `None`?

File: src/main/scala/cromwell/binding/WdlNamespace.scala (hunk 4 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: So if I understand correctly, your suggestion is to make this a coercion responsibility.  In that case, `${sep=" " File in+}` would be stored as a `CommandParameterPart` with a wdlType of `Array[File]` in which a `File` could be coerced into an `Array[File]` for that type?

If I'm understanding correctly, I think that's a pretty good solution.

Comment (mcovarr) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: Yup that's what I meant.

File: src/main/scala/cromwell/binding/TaskInput.scala (hunk 3 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/TaskInput.scala]: I suppose it could

File: src/main/scala/cromwell/binding/WdlNamespace.scala (hunk 13 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: Hmm it says that prefix is an `Option[Some[String]]` if I try that.

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/WdlNamespace.scala]: Bleh, this is one of those cases where Java bleeds in because `ast.getAttribute` can return `null`, but the return type is just `Ast`
